Implement decoding the binary form

fitzgen opened the pull request.

fitzgen committed:
Make sure to test with all features enabled in CI

fitzgen committed:
Implement decoding the Web IDL bindings custom section from binary

fitzgen committed:
Add a bunch of tests for decoding

fitzgen committed:
rustfmt

Hywan commented:
Fix #5.

Hywan reviewed (approved):
That's an impressive PR. I reviewed it carefully against the `BINARY.md` and I've found no error.

alexcrichton commented on src/binary/decode.rs:
```diff
@@ -0,0 +1,2179 @@
+    fn decode<Cx, R>(cx: &mut Cx, r: &mut R) -> Result<Self::Output, failure::Error>
+    where
+        Cx: Ids,
+        R: Read;
```
FWIW the codegen and wonkiness of `Read` I've found is pretty unfortunate historically, could this perhaps just be `&mut &[u8]` or something like that?

fitzgen replied:
Yeah, I've seen this before too. I'll see what I can do.

alexcrichton commented on src/binary/decode.rs:
```diff
@@ -0,0 +1,2179 @@
+    fn string(&mut self) -> Result<String, failure::Error>;
+}
+
+impl<R: Read> ReadExt for R {
```
I wonder if it'd be a bit cleaner to expose these methods as `Decode for TheType`? For example the `Vec` method could be `Decode for Vec<T>` which albeit not as generic may still suffice? `uleb` could be something like `Decode for u32` or something like that too.

fitzgen replied:
This would work for `u32` but not for `Vec<T>` since we don't always want to allocate a vec of the things we deserialize. I'll play with it a bit.

alexcrichton commented on src/binary/decode.rs:
```diff
@@ -0,0 +1,2179 @@
+    type Output;
+
+    /// Decode an instance of `Self` into the given `cx`.
+    fn decode<Cx, R>(cx: &mut Cx, r: &mut R) -> Result<Self::Output, failure::Error>
```
This signature is really throwing me for a spin relative to what I'd expect from a `Decode` I think. I sort of just assumed that this would return `failure::Result<Self>` and then got pretty confused at all the usage of `Ignore` and `Output = ()`. I found it somewhat difficult to see what's being inserted where. 

Would it be possible to remove the `Output` associated type and change this to just return `Self`? I think that might help making it a bit more clear for me and having one location to insert types, for example, vs multiple locations that insert types.

fitzgen replied:
Unfortunately, `Self` as the return type doesn't work so well with the arena approach. I did remove the `Cx` type param and am looking into removing the `Read` param too, which should help clean things up.

fitzgen replied:
I investigated this a bit but the hard part comes from thigns that reference other things by index (like webidl types) where we haven't yet finished constructing a whole `WebidlTypes` and its arena. This is why we sometimes decode by mutating the under-construction custom section in the cx instead of always returning the decoded value. I don't see a nice/easy way to work around this.

alexcrichton reviewed (commented):
Looks great to me! I find it kind of hard to follow the `Decode` trait but it's not really the end of the world. 

I might recommend switching to `&[u8]` forcibly though to avoid monomorphizing this entire module each time it's used, and it should also make codegen easier in rustc since dealing with generic reads tends to involve a lot of code anyway.

The tests look quite extensive though so I'd be fine merging whenever!

Hywan commented:
How do you “extract” the resulting `WebidlBindings` struct with `on_parse`? The following code does not compile:

```rust
let mut bindings;
config.on_parse(|module, ids| { bindings = decode(ids, &mut data); Ok(()) });
```

I use `on_parse` because `IndicesToIds` is required for `decode`, and this is the only place where an instance of `IndicesToIds` exists.

Edit: It's possible to do everything inside `on_parse`, but am I missing something in the API?

fitzgen committed:
decode: Remove the `Cx` type parameter

It wasn't actually ever used except as `DecodeContext`. The idea was to make
testing easier, but `DecodeContext` works just fine in the tests.

fitzgen committed:
decode: Remove the `R: Read` type parameter and work with slices directly

fitzgen committed:
decode: Add a doc comment for `Ignore`
